File: balboa-core/src/main/java/com/socrata/balboa/metrics/data/DataStoreFactory.java
package com.socrata.balboa.metrics.data;

import com.socrata.balboa.metrics.config.Configuration;
import com.socrata.balboa.metrics.config.ConfigurationException;
import com.socrata.balboa.metrics.data.impl.*;

import java.io.IOException;


public class DataStoreFactory
{

    private volatile static DataStore defaultDataStore = null;
    public static DataStore get() {
        if (defaultDataStore != null)
            return defaultDataStore;
        try {
            synchronized(DataStoreFactory.class) {
                if (defaultDataStore != null)
                    return defaultDataStore;
                defaultDataStore = get(Configuration.get());
            }
            return defaultDataStore;
        } catch (IOException e) {
            throw new ConfigurationException("Unable to determine which datastore to use because the configuration couldn't be read.", e);
        }
    }

    public static DataStore get(Configuration conf)
    {
        String datastore = (String)conf.get("balboa.datastore");

        if (datastore.equals("buffered-cassandra")) {
            return new BufferedDataStore(
                    new BadIdeasDataStore(
                            new CassandraDataStore(
                                    new CassandraQueryImpl(
                                            CassandraUtil.initializeContext(conf)))));
        }

        if (datastore.equals("cassandra"))
        {
            return new BadIdeasDataStore(
                    new CassandraDataStore(
                            new CassandraQueryImpl(
                                    CassandraUtil.initializeContext(conf))));
        }
        else
        {
            throw new ConfigurationException("Unknown datastore '" + datastore + "'.");
        }
    }
}

File: balboa-core/src/main/scala/com/socrata/balboa/metrics/data/DataStoreFactory.scala
package com.socrata.balboa.metrics.data

import java.io.IOException

import com.socrata.balboa.metrics.config.ConfigurationException
import com.socrata.balboa.metrics.data.impl._
import com.typesafe.config.{Config, ConfigFactory}

trait DataStoreFactory {
  def get: DataStore
  def get(conf: Config): DataStore
}

object DefaultDataStoreFactory extends DataStoreFactory {

  private lazy val defaultDataStore: DataStore = try {
    get(ConfigFactory.load())
  } catch {
    case ioe: IOException =>
      throw new
          ConfigurationException(
            "Unable to determine which datastore to use because the configuration couldn't be read.", ioe)
  }

  def self: DataStoreFactory = this // For convenient Java access

  def get: DataStore = defaultDataStore

  def get(conf: Config): DataStore = {
    lazy val datastore: String = conf.getString("balboa.datastore")
    lazy val bufferGranularity: Long = conf.getLong("buffer.granularity")

    datastore match {
      case "buffered-cassandra" =>
        new BufferedDataStore(
          new BadIdeasDataStore(
            new CassandraDataStore(
              new CassandraQueryImpl(
                CassandraUtil.initializeContext(conf)))), bufferGranularity = bufferGranularity)
      case "cassandra" =>
        new BadIdeasDataStore(
          new CassandraDataStore(
            new CassandraQueryImpl(
              CassandraUtil.initializeContext(conf))))
      case "raw" =>
        new CassandraDataStore(
          new CassandraQueryImpl(
            CassandraUtil.initializeContext(conf)))
      case _ =>
        throw new ConfigurationException("Unknown datastore '" + datastore + "'.")
    }
  }
}
